tensorboard/webapp/widgets/data_table/data_table_component.ts

```diff
@@ -200,4 +211,50 @@ export class DataTableComponent {
     }
     this.sortDataBy.emit({header, order: SortingOrder.ASCENDING});
   }
+
+  dragStart(header: ColumnHeaders) {
+    this.draggingHeader = header;
+
+    // This stop the end drag animation
+    document.addEventListener('dragover', function (e) {
+      e.preventDefault();
+    });
+  }
+
+  dragEnd() {
+    if (!this.draggingHeader || !this.highlightedColumn) {
+      return;
+    }
+
+    this.orderColumns.emit(
+      this.moveHeader(
+        this.headers.indexOf(this.draggingHeader),
+        this.headers.indexOf(this.highlightedColumn)
+      )
+    );
+    this.draggingHeader = undefined;
+    this.highlightedColumn = undefined;
+  }
+
+  dragEnter(header: ColumnHeaders) {
+    if (!this.draggingHeader) {
+      return;
+    }
+    if (
+      this.headers.indexOf(header) < this.headers.indexOf(this.draggingHeader)
+    ) {
+      this.highlightSide = Side.LEFT;
+    } else {
+      this.highlightSide = Side.RIGHT;
+    }
+    this.highlightedColumn = header;
+  }
+
+  moveHeader(sourceIndex: number, destinationIndex: number) {
+    let newHeaders = this.headers
+      .slice(0, sourceIndex)
+      .concat(this.headers.slice(sourceIndex + 1));
+    newHeaders.splice(destinationIndex, 0, this.headers[sourceIndex]);
+    return newHeaders;
+  }
```

Comment:
I don't really understand this function, it's just swapping the two elements, right?
```suggestion
  moveHeader(sourceIndex: number, destinationIndex: number) {
    const newHeaders = [...this.headers];
    const swap = newHeaders[sourceIndex];
    newHeaders[sourceIndex] = newHeaders[destinationIndex]
    newHeaders[destinationIndex] = swap; 
    return newHeaders;
  }
```

Reply:
As discussed this is now cleaned up to copy then use 2 splice calls. I also added comments.